### PR TITLE
Use BigQuery Read API for reading external BigLake tables

### PR DESCRIPTION
**Pull request by marcinsbd**
## Description
Continuation of the https://github.com/trinodb/trino/pull/21017

BigQuery storage APIs support reading BigLake external tables (ie external tables with a connection). But the current implementation uses views which can be expensive, because it requires Trino issuing a SQL query against BigQuery. This PR adds support to read BigLake tables directly using the storage API.

There are no behavior changes for external tables and BQ native tables - they use the view and storage APIs respectively. Added a new test for BigLake tables.

## Additional context and related issues

Fixes https://github.com/trinodb/trino/issues/21016
https://cloud.google.com/bigquery/docs/biglake-intro

## Release notes

```markdown
# BigQuery
* Improve performance when reading external BigLake tables. ({issue}`21016`)
```

### CODE REVIEW

**Comment by marcinsbd** [comment]
I will squash commits once I figured out all changes has been properly cherry-picked and rebase with master.

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java [hunk 1]
```diff
@@ -756,6 +758,43 @@ public void testBigQueryExternalTable()
...
+
+    private void assertJobCountForTable(String referencedTable, long expectedJobCount)
+    {
+        @Language("SQL")
+        String jobCountForTableQuery = """
+                     SELECT count(*) FROM region-us.INFORMATION_SCHEMA.JOBS WHERE EXISTS(
+                         SELECT * FROM UNNEST(referenced_tables) AS referenced_table
+                             WHERE referenced_table.table_id = '%s')
+                    """.formatted(referencedTable);
+         // Use assertEventually in case there are delays in the BQ information schema.
```
Comment (krvikash) [inline plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java:792]: nit: extra space

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryNamedRelationHandle.java [hunk 2]
```diff
@@ -32,20 +32,23 @@ public class BigQueryNamedRelationHandle
     private final String type;
     private final Optional<BigQueryPartitionType> partitionType;
     private final Optional<String> comment;
+    private final boolean connectionExists;
```
Comment (krvikash) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryNamedRelationHandle.java:35]: nit: rename to `connectionIdExists`

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java [hunk 3]
```diff
@@ -756,6 +758,43 @@ public void testBigQueryExternalTable()
...
+            assertQueryReturnsEmptyResult("SELECT * FROM information_schema.tables WHERE table_schema = 'test' AND table_name = '" + biglakeExternalTable + "'");
+            // BigLake tables should not run queries, since they are read directly using the storage read API.
+            assertJobCountForTable(biglakeExternalTable, 0);
+        }
+        finally {
+            onBigQuery("DROP EXTERNAL TABLE IF EXISTS test." + biglakeExternalTable);
+        }
+    }
+
+    private void assertJobCountForTable(String referencedTable, long expectedJobCount)
```
Comment (krvikash) [inline plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java:784]: we can inline this method.

**Comment by Praveen2112** [comment]
@marcinsbd Can we update the PR description

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java [hunk 4]
```diff
@@ -332,12 +333,19 @@ public ConnectorTableHandle getTableHandle(ConnectorSession session, SchemaTable
             columns.add(PARTITION_DATE.getColumnHandle());
             columns.add(PARTITION_TIME.getColumnHandle());
         }
+        Optional<String> connectionId = Optional.empty();
+        if (tableInfo.get().getDefinition() instanceof ExternalTableDefinition externalTableDefinition) {
+            if (externalTableDefinition.getConnectionId() != null && !externalTableDefinition.getConnectionId().isEmpty()) {
```
Comment (krvikash) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java:338]: can we use `isNullOrEmpty` here?

**Review by krvikash — COMMENTED** [review]
Can you update the release note?

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -644,7 +644,8 @@ jobs:
             -Dbigquery.credentials-key="${BIGQUERY_CREDENTIALS_KEY}" \
             -Dtesting.gcp-storage-bucket="${GCP_STORAGE_BUCKET}" \
             -Dtesting.bigquery-project-id="${BIGQUERY_TESTING_PROJECT_ID}" \
-            -Dtesting.bigquery-parent-project-id="${BIGQUERY_TESTING_PARENT_PROJECT_ID}"
+            -Dtesting.bigquery-parent-project-id="${BIGQUERY_TESTING_PARENT_PROJECT_ID}" \
+            -Dtesting.bigquery-connection-id="projects/sep-bq-cicd/locations/us/connections/sep-bq-cicd-biglake"
```
Comment (Praveen2112) [inline .github/workflows/ci.yml:648]: Can we inject them as a env variable ?

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java [hunk 6]
```diff
@@ -360,13 +362,14 @@ public ConnectorTableHandle getTableHandle(ConnectorSession session, SchemaTable
...
+    private static boolean useStorageApi(ConnectorSession session, String tableName, TableDefinition tableDefinition)
     {
+        TableDefinition.Type type = tableDefinition.getType();
         if (isWildcardTable(type, tableName)) {
             // Storage API doesn't support reading wildcard tables
             return false;
         }
-        if (type == EXTERNAL) {
+        if (type == EXTERNAL && !isExternalTableWithConnection(tableDefinition)) {
             // Storage API doesn't support reading external tables
```
Comment (Praveen2112) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java:373]: We can update this comment

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryNamedRelationHandle.java [hunk 7]
```diff
@@ -91,15 +100,16 @@ public boolean equals(Object o)
         // NOTE remoteTableName is not compared here because two handles differing in only remoteTableName will create ambiguity
         // TODO: Add tests for this (see TestJdbcTableHandle#testEquivalence for reference)
         return Objects.equals(schemaTableName, that.schemaTableName) &&
-                Objects.equals(type, that.type) &&
-                Objects.equals(partitionType, that.partitionType) &&
-                Objects.equals(comment, that.comment);
```
[the remark this answers is not in the document]
Comment (marcinsbd) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryNamedRelationHandle.java]: fixed

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java [hunk 8]
```diff
@@ -376,6 +379,14 @@ private static boolean useStorageApi(ConnectorSession session, String tableName,
         return true;
     }
 
+    private static boolean isBigLakeTable(TableDefinition tableDefinition)
+    {
+        if (tableDefinition instanceof ExternalTableDefinition externalTableDefinition) {
+            return !isNullOrEmpty(externalTableDefinition.getConnectionId()) && externalTableDefinition.getObjectMetadata() == null;
```
Comment (Praveen2112) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java:385]: Can we add a comment on why we are testing for null check ? Can we use `isNullOrEmpty` here ?

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/ReadSessionCreator.java [hunk 9]
```diff
@@ -140,7 +141,7 @@ private TableInfo getActualTable(
     {
         TableDefinition tableDefinition = remoteTable.getDefinition();
         TableDefinition.Type tableType = tableDefinition.getType();
-        if (tableType == TABLE || tableType == SNAPSHOT) {
+        if (tableType == TABLE || tableType == SNAPSHOT || tableType == EXTERNAL) {
```
Comment (Praveen2112) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/ReadSessionCreator.java:144]: We need to update the comments in LNo 156

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java [hunk 10]
```diff
@@ -761,12 +763,85 @@ public void testBigQueryExternalTable()
...
+    {
+        verifyHowManyTimesTableWasReferencedInJobs(objectExternalTable, 0);
+    }
+
+    private void verifyThatStreamApiWasNotUsed(String objectExternalTable)
+    {
+        verifyHowManyTimesTableWasReferencedInJobs(objectExternalTable, 1);
+    }
+
+    private void verifyHowManyTimesTableWasReferencedInJobs(String objectExternalTable, long x)
```
Comment (Praveen2112) [inline plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java:834]: This is for externalTables in general right ? How about `externalTable` and some name instead of `x`

**Comment by electrum** [comment]
/test-with-secrets sha=5c2fb2886f53965852b06d5f465b7deef089de5e

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java [hunk 11]
```diff
@@ -761,12 +763,85 @@ public void testBigQueryExternalTable()
...
+    {
+        assertThat(bigQuerySqlExecutor.executeQuery("""
+                 SELECT count(*) FROM region-us.INFORMATION_SCHEMA.JOBS WHERE EXISTS(
+                     SELECT * FROM UNNEST(referenced_tables) AS referenced_table
+                         WHERE referenced_table.table_id = '%s')
+                """.formatted(externalTable)).getValues())
+                .extracting(fieldValues -> fieldValues.getFirst().getLongValue())
+                .containsExactly(noOfReferences);
+    }
+
```
Comment (ebyhr) [inline plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java:844]: We could change this helper method to return count something like:
```java
    private long getTableReferenceCountInJob(String tableId)
    {
        return bigQuerySqlExecutor.executeQuery("""
                 SELECT count(*) FROM region-us.INFORMATION_SCHEMA.JOBS WHERE EXISTS(
                     SELECT * FROM UNNEST(referenced_tables) AS referenced_table
                         WHERE referenced_table.table_id = '%s')
                """.formatted(tableId)).streamValues()
                .map(fieldValues -> fieldValues.getFirst().getLongValue())
                .collect(onlyElement());
    }
```
We don't need to add other 2 helper methods in my opinion.

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java [hunk 12]
```diff
@@ -761,12 +763,85 @@ public void testBigQueryExternalTable()
...
+    {
+        verifyHowManyTimesTableWasReferencedInJobs(externalTable, 0);
+    }
+
+    private void verifyThatStreamApiWasNotUsed(String externalTable)
+    {
+        verifyHowManyTimesTableWasReferencedInJobs(externalTable, 1);
+    }
+
+    private void verifyHowManyTimesTableWasReferencedInJobs(String externalTable, long noOfReferences)
```
Comment (ebyhr) [inline plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java:834]:
```suggestion
    private void verifyHowManyTimesTableWasReferencedInJobs(String tableId, long expectedCount)
```

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java [hunk 13]
```diff
@@ -761,12 +763,85 @@ public void testBigQueryExternalTable()
...
+        }
+        finally {
+            onBigQuery("DROP EXTERNAL TABLE IF EXISTS test." + biglakeExternalTable);
+        }
+    }
+
+    @Test
+    public void testExternalObjectTable()
+    {
+        String objectTableConnectionId = bigQueryConnectionId;
```
Comment (ebyhr) [inline plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java:801]: No need to define `objectTableConnectionId` in my opinion.

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java [hunk 14]
```diff
@@ -761,12 +763,85 @@ public void testBigQueryExternalTable()
...
+    {
+        verifyHowManyTimesTableWasReferencedInJobs(externalTable, 1);
+    }
+
+    private void verifyHowManyTimesTableWasReferencedInJobs(String externalTable, long noOfReferences)
+    {
+        assertThat(bigQuerySqlExecutor.executeQuery("""
+                 SELECT count(*) FROM region-us.INFORMATION_SCHEMA.JOBS WHERE EXISTS(
+                     SELECT * FROM UNNEST(referenced_tables) AS referenced_table
+                         WHERE referenced_table.table_id = '%s')
```
Comment (ebyhr) [inline plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java:839]: We faced flaky issues in `assertLabelForTable` using `NFORMATION_SCHEMA.JOBS_BY_USER` table. Does this query return the latest status without delay?

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java (hunk 11 above)
Comment (marcinsbd) [reply plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java:844]: done

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java (hunk 14 above)
Comment (marcinsbd) [reply plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java]: I did not noticed locally, however I added eventual assert

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java [hunk 15]
```diff
@@ -376,6 +379,15 @@ private static boolean useStorageApi(ConnectorSession session, String tableName,
         return true;
     }
 
+    private static boolean isBigLakeTable(TableDefinition tableDefinition)
+    {
+        if (tableDefinition instanceof ExternalTableDefinition externalTableDefinition) {
+            //BigLake tables are external with connection and don't have objectMetadata
```
Comment (Praveen2112) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java:385]: connectorId

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/ReadSessionCreator.java [hunk 16]
```diff
@@ -152,7 +153,7 @@ private TableInfo getActualTable(
             // get it from the view
             return client.getCachedTable(viewExpiration, remoteTable, requiredColumns, filter);
         }
-        // Storage API doesn't support reading other table types (materialized views, external)
+        // Storage API doesn't support reading other table types (materialized views, external except BigLake tables)
```
Comment (Praveen2112) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/ReadSessionCreator.java:156]: nit: How about non-biglake external tables

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java [hunk 17]
```diff
@@ -761,12 +764,73 @@ public void testBigQueryExternalTable()
...
 
+    @Test
+    public void testBigLakeExternalTable()
+    {
+        String biglakeExternalTable = "test_biglake_external" + randomNameSuffix();
+        try {
+            onBigQuery("CREATE EXTERNAL TABLE test." + biglakeExternalTable +
+                    " WITH CONNECTION `" + bigQueryConnectionId + "`" +
+                    " OPTIONS (format = 'CSV', uris = ['gs://" + gcpStorageBucket + "/tpch/tiny/region.csv'])");
+            assertQuery("SELECT table_type FROM information_schema.tables WHERE table_schema = 'test' AND table_name = '" + biglakeExternalTable + "'", "VALUES 'BASE TABLE'");
```
Comment (Praveen2112) [inline plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java:783]: We treat everything as table in case of BigQuery right ?

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java [hunk 18]
```diff
@@ -761,12 +764,73 @@ public void testBigQueryExternalTable()
...
+        try {
+            onBigQuery("CREATE EXTERNAL TABLE test." + objectExternalTable +
+                       " WITH CONNECTION `" + bigQueryConnectionId + "`" +
+                       " OPTIONS (object_metadata = 'SIMPLE', uris = ['gs://" + gcpStorageBucket + "/tpch/tiny/region.csv'])");
+            assertQuery("SELECT table_type FROM information_schema.tables WHERE table_schema = 'test' AND table_name = '" + objectExternalTable + "'", "VALUES 'BASE TABLE'");
+
+            assertThat(query("SELECT uri FROM test." + objectExternalTable)).succeeds();
+
+            assertUpdate("DROP TABLE test." + objectExternalTable);
+            assertQueryReturnsEmptyResult("SELECT * FROM information_schema.tables WHERE table_schema = 'test' AND table_name = '" + objectExternalTable + "'");
```
Comment (Praveen2112) [inline plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java:813]: Show tables like should give us a similar response right ?

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java [hunk 19]
```diff
@@ -761,12 +764,73 @@ public void testBigQueryExternalTable()
...
+
+            // BigLake tables should not run queries, since they are read directly using the storage read API.
+            assertEventually(() -> assertThat(getTableReferenceCountInJob(objectExternalTable)).isEqualTo(1));
+        }
+        finally {
+            onBigQuery("DROP EXTERNAL TABLE IF EXISTS test." + objectExternalTable);
+        }
+    }
+
+    private long getTableReferenceCountInJob(String tableId)
```
Comment (Praveen2112) [inline plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java:823]: Should it be tableName or tableId ?

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java [hunk 20]
```diff
@@ -761,12 +764,73 @@ public void testBigQueryExternalTable()
...
+    {
+        return bigQuerySqlExecutor.executeQuery("""
+                         SELECT count(*) FROM region-us.INFORMATION_SCHEMA.JOBS WHERE EXISTS(
+                             SELECT * FROM UNNEST(referenced_tables) AS referenced_table
+                                 WHERE referenced_table.table_id = '%s')
+                        """.formatted(tableId)).streamValues()
+                .map(fieldValues -> fieldValues.getFirst().getLongValue())
```
Comment (Praveen2112) [inline plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java:830]: nit: `Iterables.getOnlyElement(bigQuery.query(QueryJobConfiguration.of(sql)).getValues()).getFirst()`

Comment (Praveen2112) [reply plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java:830]: We could also have a condition like `HAVING` clause for COUNT(*) > 0 and we could check if we have a row or not - Just an alternative approach.

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java (hunk 12 above)
Comment (Praveen2112) [reply plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java]: But this is tableName right ? Or are we passing `tableId` - cc @ebyhr

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java (hunk 14 above)
Comment (ebyhr) [reply plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java]: We shouldn't use eventual assertion if there is no delay. 

We can't recognize whether the reference count was exactly 0 or delay happened in the following assertion:
```sql
assertEventually(() -> assertThat(getTableReferenceCountInJob(biglakeExternalTable)).isEqualTo(0));
```

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java [hunk 21]
```diff
@@ -761,12 +764,73 @@ public void testBigQueryExternalTable()
...
+                       " WITH CONNECTION `" + bigQueryConnectionId + "`" +
+                       " OPTIONS (object_metadata = 'SIMPLE', uris = ['gs://" + gcpStorageBucket + "/tpch/tiny/region.csv'])");
+            assertQuery("SELECT table_type FROM information_schema.tables WHERE table_schema = 'test' AND table_name = '" + objectExternalTable + "'", "VALUES 'BASE TABLE'");
+
+            assertThat(query("SELECT uri FROM test." + objectExternalTable)).succeeds();
+
+            assertUpdate("DROP TABLE test." + objectExternalTable);
+            assertQueryReturnsEmptyResult("SELECT * FROM information_schema.tables WHERE table_schema = 'test' AND table_name = '" + objectExternalTable + "'");
+
+            // BigLake tables should not run queries, since they are read directly using the storage read API.
```
Comment (pajaks) [inline plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java:815]: This comment should be probably updated

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java [hunk 22]
```diff
@@ -761,12 +764,73 @@ public void testBigQueryExternalTable()
...
+    }
+
+    @Test
+    public void testExternalObjectTable()
+    {
+        String objectExternalTable = "test_object_external" + randomNameSuffix();
+
+        try {
+            onBigQuery("CREATE EXTERNAL TABLE test." + objectExternalTable +
+                       " WITH CONNECTION `" + bigQueryConnectionId + "`" +
```
Comment (pajaks) [inline plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java:806]: Can you add also test for external table without connection?

Comment (ebyhr) [reply plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java:806]: The existing `testBigQueryExternalTable` creates an external table without connection.

**Comment by ebyhr** [comment]
I think we should handle the following limitation:

https://cloud.google.com/bigquery/docs/biglake-intro
> The BigQuery Storage API is not available in other cloud environments, such as AWS and Azure.

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java (hunk 22 above)
Comment (pajaks) [reply plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java:806]: But there is no check if we are using storage read API there. Maybe it should be added there?

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java (hunk 15 above)
Comment (ebyhr) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java:385]: Do you mean **connection**Id?

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java (hunk 22 above)
Comment (ebyhr) [reply plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java:806]: There is 
https://github.com/trinodb/trino/pull/22974/files/f3da1b53b047bb6433628b84298baf173cac21e5#diff-a3f9fc4e76266fa0e38b1467a0ed5205aa6cd2a2b1523a4e020360803fc45a7fR767-R768. The expected behavior is REST API (not Storage API) in my understanding.

**Comment by ebyhr** [comment]
@Praveen2112 I'm not sure how BigLake Omni works in this case. How about adding another condition to `BigQueryMetadata#isBigLakeTable`? e.g.
```java
externalTableDefinition.getSourceUris().stream().allMatch(uri -> uri.startsWith("gs://")
```

**Comment by marcinsbd** [comment]
Hi @anoopj 
what would be the correct way to check if a TableDefinition describes BigLake Table (external table with connection id) or not?
is there a simple way using api to distinguish  BigLakeTable from ObjectTable and  Omni Table ?

**Comment by anoopj** [comment]
> what would be the correct way to check if a TableDefinition describes BigLake Table (external table with connection id) or not?
Not sure if I understood the question: but a BigLake table will have the connection ID in the table's [ExternalDataConfiguration](https://cloud.google.com/bigquery/docs/reference/rest/v2/tables#externaldataconfiguration)

> is there a simple way using api to distinguish BigLakeTable from ObjectTable and Omni Table ?

You can tell from the dataset region (preferred) or looking at the sourceUris as mentioned in the above comment.

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java (hunk 12 above)
Comment (marcinsbd) [reply plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java]: Let's stay with tableName

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java (hunk 14 above)
Comment (marcinsbd) [reply plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java]: The only idea that comes to my mind is periodical check with timeout. I don't like sleep in the code but if we expect delays so the sleep usage seems to be the one of the simplest solution.

**Comment by marcinsbd** [comment]
Thanks @ebyhr, @Praveen2112, @anoopj, @krvikash, @pajaks for the review and your help. AC

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java (hunk 17 above)
Comment (ebyhr) [reply plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java:800]: Right. We could remove this line.

Comment (marcinsbd) [reply plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java]: Removed.

**Review by pajaks — APPROVED** [review]

**Comment by marcinsbd** [comment]
Please let's do the another round of review @ebyhr, @Praveen2112, @krvikash.

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java [hunk 23]
```diff
@@ -1460,4 +1530,25 @@ private void onBigQuery(@Language("SQL") String sql)
     {
         bigQuerySqlExecutor.execute(sql);
     }
+
+    private static <E extends Exception> void assertConsistently(Duration timeout, Duration retryFrequency, Assert.CheckedRunnable<E> assertion)
```
Comment (Praveen2112) [inline plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java:1534]: Can we move it to `Assert` in trino

**Review by Praveen2112 — APPROVED** [review]
% comments

**Comment by Praveen2112** [comment]
@marcinsbd Can we rebase the PR

**Review by ebyhr — APPROVED** [review]
